File: static/js/components/company.js
const CompanyDashboard = {
    props: ['api', 'profile'],
    template: `
    <div>
        <h2 class="mb-4"><i class="bi bi-speedometer2 me-2"></i>Company Dashboard</h2>
        <div v-if="profile?.approval_status !== 'approved'" class="alert alert-warning">
            Your company is pending admin approval. You cannot create drives until approved.
        </div>
        <div class="row g-3 mb-4" v-if="data">
            <div class="col-md-4"><div class="card dashboard-stat shadow-sm"><div class="card-body text-center">
                <h2 class="text-primary">{{ data.drives?.length || 0 }}</h2><p class="text-muted mb-0">Placement Drives</p>
            </div></div></div>
            <div class="col-md-4"><div class="card dashboard-stat shadow-sm"><div class="card-body text-center">
                <h2 class="text-success">{{ data.total_applications || 0 }}</h2><p class="text-muted mb-0">Total Applicants</p>
            </div></div></div>
            <div class="col-md-4"><div class="card dashboard-stat shadow-sm"><div class="card-body text-center">
                <h3 class="text-info">{{ data.company?.company_name }}</h3><p class="text-muted mb-0">Your Company</p>
            </div></div></div>
        </div>
        <h4 class="mb-3">Your Drives</h4>
        <div class="table-responsive card shadow-sm" v-if="data">
            <table class="table table-hover mb-0">
                <thead class="table-light"><tr><th>Title</th><th>Status</th><th>Applicants</th><th>Deadline</th></tr></thead>
                <tbody>
                    <tr v-for="d in data.drives" :key="d.id">
                        <td>{{ d.title }}</td>
                        <td><span class="badge" :class="'bg-' + statusBadge(d.status)">{{ d.status }}</span></td>
                        <td>{{ d.applicant_count }}</td>
                        <td>{{ d.application_deadline ? new Date(d.application_deadline).toLocaleDateString() : 'N/A' }}</td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>`,
    data() { return { data: null }; },
    methods: { statusBadge },
    async mounted() { this.data = await this.api.companyDashboard(); }
};

const CompanyProfile = {
    props: ['api', 'profile'],
    emits: ['updated'],
    template: `
    <div>
        <h2 class="mb-4"><i class="bi bi-building me-2"></i>Company Profile</h2>
        <div class="card shadow-sm"><div class="card-body p-4">
            <form @submit.prevent="save">
                <div class="row">
                    <div class="col-md-6 mb-3"><label class="form-label">Company Name</label>
                        <input type="text" class="form-control" v-model="form.company_name" required></div>
                    <div class="col-md-6 mb-3"><label class="form-label">Industry</label>
                        <input type="text" class="form-control" v-model="form.industry"></div>
                    <div class="col-md-6 mb-3"><label class="form-label">Location</label>
                        <input type="text" class="form-control" v-model="form.location"></div>
                    <div class="col-md-6 mb-3"><label class="form-label">Website</label>
                        <input type="url" class="form-control" v-model="form.website"></div>
                    <div class="col-md-6 mb-3"><label class="form-label">HR Contact</label>
                        <input type="text" class="form-control" v-model="form.hr_contact"></div>
                    <div class="col-md-6 mb-3"><label class="form-label">HR Email</label>
                        <input type="email" class="form-control" v-model="form.hr_email"></div>
                    <div class="col-12 mb-3"><label class="form-label">Description</label>
                        <textarea class="form-control" v-model="form.description" rows="3"></textarea></div>
                </div>
                <div v-if="msg" class="alert alert-success">{{ msg }}</div>
                <button type="submit" class="btn btn-primary">Save Profile</button>
            </form>
        </div></div>
    </div>`,
    data() { return { form: { ...this.profile }, msg: '' }; },
    methods: {
        async save() {
            const res = await this.api.updateCompanyProfile(this.form);
            this.msg = res.message;
            this.$emit('updated', res.company);
        }
    }
};

const CompanyDrives = {
    props: ['api'],
    template: `
    <div>
        <div class="d-flex justify-content-between align-items-center mb-4">
            <h2><i class="bi bi-briefcase me-2"></i>Placement Drives</h2>
            <button class="btn btn-primary" @click="showForm = !showForm">
                <i class="bi bi-plus-lg me-1"></i>Create Drive
            </button>
        </div>
        <div v-if="showForm" class="card shadow-sm mb-4"><div class="card-body p-4">
            <h5>New Placement Drive</h5>
            <form @submit.prevent="create">
                <div class="row">
                    <div class="col-md-6 mb-3"><label class="form-label">Job Title *</label>
                        <input type="text" class="form-control" v-model="form.title" required></div>
                    <div class="col-md-6 mb-3"><label class="form-label">Salary</label>
                        <input type="text" class="form-control" v-model="form.salary" placeholder="12 LPA"></div>
                    <div class="col-12 mb-3"><label class="form-label">Description</label>
                        <textarea class="form-control" v-model="form.description" rows="2"></textarea></div>
                    <div class="col-md-6 mb-3"><label class="form-label">Skills Required</label>
                        <input type="text" class="form-control" v-model="form.skills_required"></div>
                    <div class="col-md-6 mb-3"><label class="form-label">Experience</label>
                        <input type="text" class="form-control" v-model="form.experience_required"></div>
                    <div class="col-md-4 mb-3"><label class="form-label">Eligible Branch</label>
                        <input type="text" class="form-control" v-model="form.eligibility_branch" placeholder="CSE, ECE"></div>
                    <div class="col-md-4 mb-3"><label class="form-label">Min CGPA</label>
                        <input type="number" step="0.1" class="form-control" v-model.number="form.eligibility_cgpa"></div>
                    <div class="col-md-4 mb-3"><label class="form-label">Eligible Year</label>
                        <input type="number" class="form-control" v-model.number="form.eligibility_year"></div>
                    <div class="col-md-6 mb-3"><label class="form-label">Application Deadline</label>
                        <input type="datetime-local" class="form-control" v-model="form.application_deadline"></div>
                </div>
                <div v-if="error" class="alert alert-danger">{{ error }}</div>
                <button type="submit" class="btn btn-success">Submit for Approval</button>
            </form>
        </div></div>
        <div class="table-responsive card shadow-sm">
            <table class="table table-hover mb-0">
                <thead class="table-light"><tr>
                    <th>Title</th><th>Status</th><th>Deadline</th><th>Actions</th>
                </tr></thead>
                <tbody>
                    <tr v-for="d in drives" :key="d.id">
                        <td>{{ d.title }}</td>
                        <td><span class="badge" :class="'bg-' + statusBadge(d.status)">{{ d.status }}</span></td>
                        <td>{{ d.application_deadline ? new Date(d.application_deadline).toLocaleDateString() : 'N/A' }}</td>
                        <td>
                            <button v-if="['approved','active'].includes(d.status)" class="btn btn-sm btn-warning me-1" @click="closeDrive(d.id)">Close</button>
                            <button v-if="d.status==='approved'" class="btn btn-sm btn-success" @click="activateDrive(d.id)">Activate</button>
                        </td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>`,
    data() {
        return {
            drives: [], showForm: false, error: '',
            form: { title: '', description: '', skills_required: '', experience_required: '',
                    salary: '', eligibility_branch: '', eligibility_cgpa: null, eligibility_year: null,
                    application_deadline: '' }
        };
    },
    methods: {
        statusBadge,
        async load() { this.drives = await this.api.companyDrives(); },
        async create() {
            try {
                const data = { ...this.form };
                if (data.application_deadline) data.application_deadline = new Date(data.application_deadline).toISOString();
                await this.api.createDrive(data);
                this.showForm = false;
                this.form = { title: '', description: '', skills_required: '', experience_required: '',
                    salary: '', eligibility_branch: '', eligibility_cgpa: null, eligibility_year: null, application_deadline: '' };
                await this.load();
            } catch (e) { this.error = e.message; }
        },
        async closeDrive(id) { await this.api.updateDriveStatus(id, 'closed'); await this.load(); },
        async activateDrive(id) { await this.api.updateDriveStatus(id, 'active'); await this.load(); },
    },
    mounted() { this.load(); }
};

const CompanyApplications = {
    props: ['api'],
    template: `
    <div>
        <h2 class="mb-4"><i class="bi bi-people me-2"></i>Manage Applications</h2>
        <div class="mb-3">
            <select class="form-select" style="max-width:300px" v-model="selectedDrive" @change="loadApps">
                <option value="">Select a drive...</option>
                <option v-for="d in drives" :key="d.id" :value="d.id">{{ d.title }}</option>
            </select>
        </div>
        <div v-if="apps.length" class="table-responsive card shadow-sm">
            <table class="table table-hover mb-0">
                <thead class="table-light"><tr>
                    <th>Student</th><th>Branch</th><th>CGPA</th><th>Status</th><th>Actions</th>
                </tr></thead>
                <tbody>
                    <tr v-for="a in apps" :key="a.id">
                        <td>{{ a.student?.full_name }}</td>
                        <td>{{ a.student?.branch }}</td>
                        <td>{{ a.student?.cgpa }}</td>
                        <td><span class="badge" :class="'bg-' + statusBadge(a.status)">{{ a.status }}</span></td>
                        <td>
                            <select class="form-select form-select-sm" style="width:auto;display:inline-block"
                                    :value="a.status" @change="updateStatus(a.id, $event.target.value)">
                                <option value="applied">Applied</option>
                                <option value="shortlisted">Shortlisted</option>
                                <option value="interview">Interview</option>
                                <option value="offer">Offer</option>
                                <option value="selected">Selected</option>
                                <option value="rejected">Rejected</option>
                                <option value="placed">Placed</option>
                            </select>
                        </td>
                    </tr>
                </tbody>
            </table>
        </div>
        <p v-else-if="selectedDrive" class="text-muted">No applications for this drive.</p>
    </div>`,
    data() { return { drives: [], selectedDrive: '', apps: [] }; },
    methods: {
        statusBadge,
        async load() { this.drives = await this.api.companyDrives(); },
        async loadApps() {
            if (!this.selectedDrive) { this.apps = []; return; }
            this.apps = await this.api.driveApplications(this.selectedDrive);
        },
        async updateStatus(id, status) {
            await this.api.updateApplicationStatus(id, { status });
            await this.loadApps();
        }
    },
    mounted() { this.load(); }
};

File: routes/company.py
from datetime import datetime, timezone
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import db
from models.models import User, Company, JobPosition, Application, Placement, Student
from utils.auth_helpers import role_required
from utils.cache import cache_delete_pattern

company_bp = Blueprint('company', __name__)


def _get_company():
    user_id = int(get_jwt_identity())
    user = User.query.get(user_id)
    if not user or not user.company:
        return None
    return user.company


@company_bp.route('/dashboard', methods=['GET'])
@jwt_required()
@role_required('company')
def dashboard():
    company = _get_company()
    if not company:
        return jsonify({'error': 'Company profile not found'}), 404
    if company.approval_status != 'approved':
        return jsonify({'error': 'Company not approved yet'}), 403

    drives = company.job_positions.all()
    drive_stats = []
    for d in drives:
        drive_stats.append({
            **d.to_dict(),
            'applicant_count': d.applications.count(),
        })

    return jsonify({
        'company': company.to_dict(),
        'drives': drive_stats,
        'total_applications': sum(d.applications.count() for d in drives),
    })


@company_bp.route('/profile', methods=['GET', 'PUT'])
@jwt_required()
@role_required('company')
def profile():
    company = _get_company()
    if not company:
        return jsonify({'error': 'Company profile not found'}), 404

    if request.method == 'GET':
        return jsonify(company.to_dict())

    data = request.get_json() or {}
    for field in ['company_name', 'industry', 'location', 'website', 'hr_contact', 'hr_email', 'description']:
        if field in data:
            setattr(company, field, data[field])
    db.session.commit()
    return jsonify({'message': 'Profile updated', 'company': company.to_dict()})


@company_bp.route('/drives', methods=['GET', 'POST'])
@jwt_required()
@role_required('company')
def drives():
    company = _get_company()
    if not company:
        return jsonify({'error': 'Company profile not found'}), 404
    if company.approval_status != 'approved':
        return jsonify({'error': 'Company must be approved to create drives'}), 403

    if request.method == 'GET':
        status = request.args.get('status')
        query = company.job_positions
        if status:
            query = query.filter_by(status=status)
        return jsonify([d.to_dict() for d in query.order_by(JobPosition.created_at.desc()).all()])

    data = request.get_json() or {}
    if not data.get('title'):
        return jsonify({'error': 'Job title is required'}), 400

    deadline = None
    if data.get('application_deadline'):
        deadline = datetime.fromisoformat(data['application_deadline'].replace('Z', '+00:00'))

    drive = JobPosition(
        company_id=company.id,
        title=data['title'],
        description=data.get('description', ''),
        skills_required=data.get('skills_required', ''),
        experience_required=data.get('experience_required', ''),
        salary=data.get('salary', ''),
        benefits=data.get('benefits', ''),
        eligibility_branch=data.get('eligibility_branch', ''),
        eligibility_cgpa=data.get('eligibility_cgpa'),
        eligibility_year=data.get('eligibility_year'),
        application_deadline=deadline,
        status='pending',
    )
    db.session.add(drive)
    db.session.commit()
    cache_delete_pattern('jobs:*')
    cache_delete_pattern('admin:*')
    return jsonify({'message': 'Drive created, pending admin approval', 'drive': drive.to_dict()}), 201


@company_bp.route('/drives/<int:drive_id>/status', methods=['PUT'])
@jwt_required()
@role_required('company')
def update_drive_status(drive_id):
    company = _get_company()
    drive = JobPosition.query.filter_by(id=drive_id, company_id=company.id).first_or_404()
    data = request.get_json() or {}
    new_status = data.get('status')
    if new_status not in ('active', 'closed'):
        return jsonify({'error': 'Invalid status. Use active or closed'}), 400
    if drive.status not in ('approved', 'active', 'closed'):
        return jsonify({'error': 'Drive must be approved first'}), 400
    drive.status = new_status
    db.session.commit()
    cache_delete_pattern('jobs:*')
    return jsonify({'message': 'Drive status updated', 'drive': drive.to_dict()})


@company_bp.route('/drives/<int:drive_id>/applications', methods=['GET'])
@jwt_required()
@role_required('company')
def drive_applications(drive_id):
    company = _get_company()
    drive = JobPosition.query.filter_by(id=drive_id, company_id=company.id).first_or_404()
    apps = drive.applications.order_by(Application.application_date.desc()).all()
    return jsonify([a.to_dict() for a in apps])


@company_bp.route('/applications/<int:app_id>/status', methods=['PUT'])
@jwt_required()
@role_required('company')
def update_application_status(app_id):
    company = _get_company()
    app = Application.query.get_or_404(app_id)
    if app.job_position.company_id != company.id:
        return jsonify({'error': 'Access denied'}), 403

    data = request.get_json() or {}
    status = data.get('status')
    valid = ('applied', 'shortlisted', 'interview', 'offer', 'selected', 'rejected', 'placed')
    if status not in valid:
        return jsonify({'error': f'Invalid status. Use one of: {valid}'}), 400

    app.status = status
    app.feedback = data.get('feedback', app.feedback)
    if data.get('interview_date'):
        app.interview_date = datetime.fromisoformat(data['interview_date'].replace('Z', '+00:00'))
    if data.get('interview_location'):
        app.interview_location = data['interview_location']
    app.updated_at = datetime.now(timezone.utc)

    if status in ('selected', 'placed'):
        existing = Placement.query.filter_by(
            student_id=app.student_id, job_id=app.job_id
        ).first()
        if not existing:
            placement = Placement(
                student_id=app.student_id,
                company_id=company.id,
                job_id=app.job_id,
                position=app.job_position.title,
                salary=app.job_position.salary,
            )
            db.session.add(placement)

    db.session.commit()
    cache_delete_pattern('admin:*')
    return jsonify({'message': 'Application updated', 'application': app.to_dict()})


@company_bp.route('/students/<int:student_id>', methods=['GET'])
@jwt_required()
@role_required('company')
def view_student(student_id):
    student = Student.query.get_or_404(student_id)
    return jsonify(student.to_dict())
